### ISSUE TITLE
Creating a SmallVec for every pixel is slow

### ISSUE THREAD

**Issue by Shnatsel** [johannesvollmer/exrs#172]
I've been looking into the TODO in the README:

`TODO profile if smallvec is really an improvement!`

`SmallVec` is all over the place, so it's rather difficult to remove it and benchmark the before/after. I've used Rust's [instrumentation-based coverage](https://doc.rust-lang.org/stable/rustc/instrument-coverage.html) to find out how frequently a `SmallVec` is constructed. I've profiled the `7_crop_alpha_any_image` example. Here's the report in HTML (trimmed so that only covered lines are shown): [results_gt1.html.gz](https://github.com/johannesvollmer/exrs/files/10310525/results_gt1.html.gz)

I've found that an instance of this type is constructed for every pixel:

https://github.com/johannesvollmer/exrs/blob/49fece069f9bc1e21b252fe0e98311867c38f5d3/src/image/mod.rs#L493

Profiling shows that 55% of the execution time is spent just creating `SmallVec`s for every pixel in the crop function: https://share.firefox.dev/3WQyWUR

This will be even worse on images with more than 8 channels, where a heap allocation will be performed for every pixel.

It seems that the approach of creating a SmallVec for every pixel is fundamentally slow. A much more performant way to implement this would be returning a slice of the original data, without constructing an owned data structure.

**Comment by Shnatsel** [comment]
The [`sample_vec_at`](https://github.com/johannesvollmer/exrs/blob/49fece069f9bc1e21b252fe0e98311867c38f5d3/src/image/mod.rs#L498-L501) function is to blame.

**Comment by johannesvollmer** [comment]
thanks for investing the time! great observations :)

**Comment by johannesvollmer** [comment]
the most important question is: is smallvec slower than a standard vec? because usually, slices are used where possible already. most of the time, smallvec is only used where sliced cannot be used unfortunately

**Comment by Shnatsel** [comment]
Instead of creating a SmallVec for every pixel, you can have the caller provide a `&mut [u8]` to which the pixel will be written. This allows for a single allocation to be reused.

```
    pub fn output_sample_at(&self, position: Vec2<usize>, output_buffer: &mut [u8]) -> FlatSamplesPixel {
        output_buffer.copy_from_slice(self.samples_at(position))
    }
```

And the usage would look roughly like this:

```rust
// pre-allocate a buffer once; a Vec is faster in this case
let mut buffer = vec![0; bytes_per_pixel];
// iterate over all pixels in an image
// note: iteration order (rows first or columns first) 
// must match the image layout for performance
for x in 0..img_width {
    for y in 0..img_height {
        image.output_pixel_at(x, y, &mut buffer);
        // the pixel was written to the buffer
        process_pixel(buffer); // whatever we want to do with it
    }
}
```

This API also allows for cheaper transformation of images without copying the data to/from the stack. You could allocate a large Vec for the output image, iterate over it with `slice.chunks_exact(bytes_per_pixel)` and have the iterator write directly to the output Vec, then make any further transformations in place if needed. This is great for images where only some parts of them have to be transformed, and the transformation preserves their size.

**Comment by johannesvollmer** [comment]
I'm a little bit hesitant with this approach. I fear it might not even be possible. For example, some storages do not store the pixel in one location, but in multiple locations: They store each channel in one vector, such that a pixels green value is in a different buffer than the red value. 

That's why I fear there could be no way to have one slice contain the whole pixel. But what might be possible is to return an `impl Iterator<Sample>` instead of a `SmallVec<Sample>`. Something we could try :)

**Comment by Shnatsel** [comment]
> Also, there is a function that avoids the allocation of the smallvec, returning an iterator instead. The iterator can be written into a pre allocated vector. Maybe this is something the crop function needs to use instead of sample_vec_at? This function might currently not exist in the general image trait, but maybe we can make that happen?

Yes, I believe this would resolve the performance pitfall. The function returning a SmallVec is just a convenience wrapper around this one, correct? If so, I suggest exposing this function everywhere and marking the one returning SmallVec as deprecated.

---

Some thoughts on the `&mut [u8]` for output are included below, but they're less relevant if the iterator is available (it's only a convenience wrapper anyway).

> The content of the byte slice can not be interpreted without knowing about the channel layout of the particular image.

I'm not sure I follow. The proposed function will write out the exact same sequence of bytes as the one returning a SmallVec, the difference is that the allocation is controlled by the caller.

The current calls look something like this (ignoring the `Vec2` type):

```rust
let pixel = img.samples_at(x, y);
process_pixel(&mut pixel);
```

While a call to the proposed interface would be:

```rust
let buffer = vec![0; sample_size];
output_samples_at(x, y, &mut buffer);
process_pixel(&mut buffer);
```

Except nobody ever looks up just one pixel, so calling `output_samples_at` in a loop would be dramatically more efficient.

The only difference is that the required length of the buffer needs to be known before calling `output_sample_at`.

**Comment by johannesvollmer** [comment]
I would like to implement a prototypal refactoring to show what I mean :) but maybe I cannot get it done today

**Comment by johannesvollmer** [comment]
@Shnatsel would you mind summarizing your approach to profiling? And how you generated those interactive reports? What software did you use? I'm interested in profiling this library more

**Comment by Shnatsel** [comment]
I have an entire article written about this, but it's currently not published. Drop me a line at the email in my Github profile and I'll share the draft with you.
